### PR TITLE
feat(schema): add layered global overrides

### LINKED ISSUE #1687: Proposal: layered global schema overrides that continue receiving packaged updates
# Proposal: layered global schema overrides that continue receiving packaged updates

Implementation: #1686

## Problem

OpenSpec supports global schema customization today by placing a complete `schema.yaml` and its templates in the user data directory. That is useful when a user wants to own an entire workflow, but it also permanently shadows the packaged schema.

My use case is smaller and additive: I want to keep the maintained `spec-driven` workflow while adding personal guidance to an existing artifact such as `tasks`, and perhaps replacing one template. With the current complete-replacement model, making that small change means copying the entire schema bundle and becoming responsible for manually tracking every later improvement to packaged instructions, dependencies, and templates.

This creates an unfortunate choice:

1. Keep the packaged schema and cannot express reusable personal schema guidance globally.
2. Fork the complete schema globally and stop receiving upstream workflow improvements automatically.
3. Repeat similar rules in each project, even when they are personal and apply across projects.

The desired behavior is: inherit the packaged workflow by default, explicitly customize only selected fields or templates, and continue receiving package updates everywhere else.

## Proposed shape

Add a second user-level customization mode:

```text
${XDG_DATA_HOME}/openspec/schemas/<name>/schema.override.yaml
```

The existing user file keeps its current meaning:

```text
${XDG_DATA_HOME}/openspec/schemas/<name>/schema.yaml
```

- `schema.yaml` is a complete, self-contained replacement.
- `schema.override.yaml` is a small layer applied only to the packaged schema.

For example:

```yaml
patchVersion: 1

artifacts:
  tasks:
    instruction:
      append: |
        Additional rules:
        - Include verification commands in every task group.
        - Mention the affected package in each task.
```

If a later OpenSpec version updates the packaged `tasks` instruction, the new packaged text remains active and the personal rules are still appended. A field explicitly replaced by the override remains user-owned.

The proposed command:

```bash
openspec schema override spec-driven
```

creates a valid no-op override without copying the packaged schema or templates.

## Resolution model

Schema precedence would be:

1. Complete project `schema.yaml`
2. Complete user `schema.yaml`
3. Packaged `schema.yaml` plus user `schema.override.yaml`
4. Packaged schema unchanged

Project schemas remain highest priority because they represent version-controlled team intent. A layered user override is personal and should not modify an explicit project workflow.

The two user modes are intentionally mutually exclusive. If `schema.yaml` and `schema.override.yaml` exist in the same active user schema directory, resolution should report a conflict rather than silently selecting one. An override without a packaged schema of the same name is also invalid; custom workflows continue to use complete schemas.

## Explicit merge operations

I do not propose generic YAML deep-merge behavior. Arrays and text are too easy to interpret differently, and an implicit merge can silently change meaning when the packaged schema evolves.

Instead, the override has a strict, versioned format with field-specific operations:

- Instruction text: `prepend`, `append`, or `replace`.
- Dependency arrays: `add`, `remove`, or `replace`.
- Artifact `description`, `generates`, and `template`: direct replacement.

`prepend` and `append` may be combined. `replace` is mutually exclusive with additive operations. For arrays, removals preserve the relative order of retained values and additions are appended in declaration order.

Unknown fields, unknown artifact IDs, duplicate dependencies, invalid removals, operation conflicts, invalid references, and cycles should fail validation. The final composed result should pass the same complete-schema validation as an ordinary schema.

Overlays would not add, remove, or reorder artifacts. Those are structural workflow changes and are better represented by a complete schema fork.

## Template behavior

The user directory may contain only selected whole-file template replacements:

```text
schemas/spec-driven/
├── schema.override.yaml
└── templates/
    └── tasks.md
```

For a composed schema, each referenced template resolves independently:

1. User override template
2. Packaged template fallback

In this example, `tasks.md` is user-owned, while proposal, specs, and design templates continue receiving packaged updates. Template Markdown is not merged.

Complete project and user schemas remain self-contained and do not gain package fallback, preserving current behavior.

## Visibility and validation

Layering should be visible rather than magical. The proposal makes the effective sources available through existing commands:

- `schema which` reports package plus user overlay and both paths.
- `schema validate` validates the override, composed schema, and effective templates.
- `schemas` reports effective composed metadata.
- `templates` reports whether each concrete template came from the user or package directory.
- `schema fork` materializes the effective schema and templates into a complete project bundle when structural customization is eventually needed.

Runtime loading and diagnostics should consume one centralized source descriptor so they cannot disagree about precedence or template selection.

## Backward compatibility and migration

This is additive. Existing project schemas, complete global replacements, and package-only resolution retain their behavior.

Migration from a complete global replacement is manual and explicit: translate only the desired field changes into `schema.override.yaml`, keep only intentionally replaced templates, remove the complete user `schema.yaml`, then validate and inspect the effective sources.

Rollback is simply removing `schema.override.yaml` and optional user templates; the packaged schema immediately becomes active unchanged.

## Alternatives considered

- **Continue using complete global forks.** This provides full control but is the source of the update-drift problem for small additive customizations.
- **Use project rules everywhere.** Project rules remain useful for team-owned repository constraints, but they do not represent reusable personal schema guidance across projects and do not customize template selection.
- **Treat a partial `schema.yaml` as an overlay.** This would silently reinterpret an existing filename and make complete and partial ownership ambiguous.
- **Use generic deep merge or YAML merge keys.** This leaves text and array behavior unclear and makes package evolution harder to reason about.
- **Add arbitrary schema inheritance.** This is much broader than the current use case and introduces multi-level resolution and conflict questions that are unnecessary for a package-plus-user layer.

## Non-goals

- Project-level overlay files; projects continue to use complete version-controlled schemas and project rules.
- Arbitrary `extends` chains between schemas.
- Adding, removing, or reordering artifacts through an overlay.
- Merging Markdown template contents.
- Automatically converting or rebasing existing complete user schemas.

## Questions

1. Does the distinction between complete ownership (`schema.yaml`) and layered customization (`schema.override.yaml`) fit the intended schema model?
2. Are explicit text and collection operations preferable to a generic merge format?
3. Should layered customization remain limited to packaged schemas, or is there a concrete need for project/user schema inheritance later?

### PR DESCRIPTION
**Pull request by mehdishahdoost**
## Motivation

OpenSpec already supports global schema customization, but the current mechanism is a complete replacement: users copy `schema.yaml` and every referenced template into their user data directory. From that point onward, the copied bundle shadows the packaged schema entirely.

That works when someone wants to own a whole workflow, but it is costly for the more common additive case. For example, I want to keep the maintained `spec-driven` workflow and add a few personal rules to the `tasks` instruction. With a complete replacement, I also become responsible for manually tracking every later improvement OpenSpec makes to built-in instructions, dependencies, validation behavior, and templates.

This PR adds a second, explicitly layered customization mode so small personal changes can coexist with upstream schema evolution.

## Proposed solution

A user can now create:

```text
${XDG_DATA_HOME}/openspec/schemas/<name>/schema.override.yaml
```

The packaged `schema.yaml` remains the base. OpenSpec applies the user overlay to that base at resolution time and then validates the complete effective schema using the existing schema validation rules.

For example:

```yaml
patchVersion: 1

artifacts:
  tasks:
    instruction:
      append: |
        Additional rules:
        - Include verification commands in every task group.
        - Mention the affected package in each task.
```

If a later OpenSpec release updates the packaged task instruction, the updated instruction is still used and the personal guidance remains appended. Fields that the overlay explicitly replaces remain user-owned.

The new command:

```bash
openspec schema override spec-driven
```

creates a valid no-op overlay without copying the packaged schema or templates.

## Merge model

The overlay format is deliberately strict and versioned with `patchVersion: 1`. It uses explicit operations instead of generic YAML merge semantics:

- Instruction text supports `prepend`, `append`, or `replace`. `prepend` and `append` may be combined; `replace` is exclusive.
- Dependency arrays support `add`, `remove`, or `replace`. Removal preserves the order of remaining values and additions are appended in declaration order.
- Artifact `description`, `generates`, and `template` values are direct replacements.
- Unknown keys, unknown artifact IDs, duplicate dependencies, invalid removals, operation conflicts, invalid references, and dependency cycles are rejected.

Overlays intentionally cannot add, remove, or reorder artifacts. Those are structural workflow changes and remain a use case for a complete schema fork.

## Template behavior

An overlay may include only the templates the user wants to replace:

```text
schemas/spec-driven/
├── schema.override.yaml
└── templates/
    └── tasks.md
```

For a composed schema, each template resolves independently:

1. User override template
2. Packaged template fallback

In this example, `tasks.md` is user-owned while proposal, specs, and design templates continue to receive packaged updates. Template content is not merged; a user template is a whole-file replacement.

Complete project and user schemas remain self-contained and do not gain package fallback.

## Resolution and compatibility

Schema precedence becomes:

1. Complete project `schema.yaml`
2. Complete user `schema.yaml`
3. Packaged `schema.yaml` plus user `schema.override.yaml`
4. Packaged schema without an overlay

This preserves existing behavior for project schemas, complete global replacements, and package-only users. A project schema remains authoritative because it represents version-controlled team intent.

The two user customization modes are mutually exclusive. If `schema.yaml` and `schema.override.yaml` exist in the same active user schema directory, OpenSpec reports a conflict instead of silently choosing one. An overlay without a packaged base is also rejected.

## Visibility and tooling

The same centralized source descriptor now drives runtime loading and diagnostics, so the CLI cannot report one source while using another. This PR updates:

- `schema which` and `schema which --all` to show package-plus-overlay composition
- `schema validate` to validate the overlay, effective schema, and layered templates
- `schemas` to report effective composed metadata
- `templates` to report the concrete user or package source for every template
- `schema fork` to materialize a composed schema and all effective templates into a self-contained project bundle

All template candidates retain the existing canonical path-containment checks.

## Scope

This PR does not add project-level overlays, arbitrary schema inheritance, Markdown merging, artifact insertion/removal/reordering, or automatic migration of complete user schemas. Existing complete replacements remain available for users who intentionally want full ownership.

## Testing

- `npm run lint`
- `node build.js`
- 262 focused and regression tests passed
- `openspec validate add-global-schema-overlays --strict`
- Full local suite: 3,992 of 3,993 tests passed. The remaining environment-specific workset test launched the installed Claude binary instead of its intentionally broken fixture; it is unrelated to schema resolution.

Closes #1687

## Summary by CodeRabbit

* **New Features**
  * Added layered user overrides for packaged schemas, including field, instruction, and dependency customizations.
  * Added `schema override` with JSON output and forced replacement options.
  * Schema validation, listing, templates, instructions, and forking now report and use effective schema sources.
  * Added safeguards for conflicting, invalid, incomplete, or unsafe schema and template configurations.

* **Documentation**
  * Expanded guidance for creating, customizing, validating, troubleshooting, and resolving schema overrides.
  * Corrected schema command synopsis formatting.

### CODE REVIEW

File: openspec/changes/add-global-schema-overlays/design.md [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+Text operations support:
+
+```yaml
+instruction:
+  prepend: text before the packaged instruction
+  append: text after the packaged instruction
+```
+
+`prepend` and `append` may be used together. Non-empty segments are joined with one blank line. `replace` is mutually exclusive with both:
+
```
[the remark this answers is not in the document]
Comment (mehdishahdoost) [reply openspec/changes/add-global-schema-overlays/design.md]: Addressed in b67442f: prepend, base, and append segments now drop whitespace-only boundary lines, omit empty segments, preserve internal whitespace, and join with exactly one blank line. The design, spec scenario, and unit test were updated; replace remains exact.

File: docs/cli.md [hunk 2]
```diff
@@ -1019,6 +1021,39 @@ openspec schema fork <source> [name] [options]
...
+
+---
+
+### `openspec schema override`
+
+Create a layered user override for a packaged schema. Unlike `schema fork`, this command does not copy `schema.yaml` or its templates; packaged updates remain active for fields and templates you do not override.
+
+```
+openspec schema override <name> [options]
+```
```
[the remark this answers is not in the document]
Comment (mehdishahdoost) [reply docs/cli.md]: Addressed in b67442f: the new usage and output fences now use the text language identifier, and the JSON docs describe the conditional projectSchemaTakesPrecedence field.

File: src/commands/schema.ts [hunk 3]
```diff
@@ -95,24 +108,37 @@ function getSchemaResolution(
...
-
-  const active = existingLocations[0];
-  const shadows = existingLocations.slice(1).map((loc) => ({
+  const activeIndex = existingLocations.findIndex(
+    (location) => location.source === sources.base.source && location.path === sources.base.dir
+  );
+  const shadows = existingLocations.slice(activeIndex + 1).map((loc) => ({
     source: loc.source,
     path: loc.path,
   }));
```
[the remark this answers is not in the document]
Comment (mehdishahdoost) [reply src/commands/schema.ts]: Addressed in b67442f: getSchemaResolution normalizes .yaml and .yml once for source, location, and overlay lookup, returns the normalized name, and treats a missing active location as an invariant failure. A .yaml command regression test was added.

File: src/core/validation/validator.ts [hunk 4]
```diff
@@ -474,13 +474,10 @@ export class Validator {
...
+      const schemaSources = resolveSchemaSources(schemaName, projectRoot);
       if (
         schemaName !== 'spec-driven' ||
-        schemaDir === null ||
-        FileSystemUtils.canonicalizeExistingPath(schemaDir) !==
-          FileSystemUtils.canonicalizeExistingPath(builtInSchemaDir)
+        schemaSources?.mode !== 'package'
       ) {
         return [];
       }
```
[the remark this answers is not in the document]
Comment (mehdishahdoost) [reply src/core/validation/validator.ts]: Addressed in b67442f by restoring the built-in schema identity gate while leaving task-file selection to resolveTaskFilesForChange, which resolves the effective overlaid tracking configuration. The new CLI regression confirms instruction-only overlays retain all three numbering warnings.

File: openspec/changes/add-global-schema-overlays/specs/schema-validate-command/spec.md [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+
+- **GIVEN** a referenced template is absent from the user overlay directory but present in the packaged directory
+- **WHEN** the user validates the composed schema
+- **THEN** template validation SHALL succeed using the packaged template
+
+#### Scenario: User replacement and overlay conflict
+
+- **GIVEN** the user schema directory contains both `schema.yaml` and `schema.override.yaml`
+- **WHEN** the user validates that schema
+- **THEN** validation SHALL fail and explain the two mutually exclusive customization modes
```
[the remark this answers is not in the document]
Comment (mehdishahdoost) [reply openspec/changes/add-global-schema-overlays/specs/schema-validate-command/spec.md:40]: Addressed in b67442f: the scenario now limits the user replacement and overlay conflict to cases with no higher-priority project schema, with a command regression test confirming an active project schema still validates.

File: test/commands/schema-overlay.test.ts [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+      'spec-driven',
+      'schema.override.yaml'
+    );
+  }
+
+  function lastJsonLog(): Record<string, unknown> {
+    const value = consoleLogSpy.mock.calls.at(-1)?.[0];
+    expect(typeof value).toBe('string');
+    return JSON.parse(value as string) as Record<string, unknown>;
+  }
```
[the remark this answers is not in the document]
Comment (mehdishahdoost) [reply test/commands/schema-overlay.test.ts:108]: Partially adopted in b67442f: both temp roots are canonicalized, concrete returned template paths are asserted canonically, and an escaping template symlink is rejected. I intentionally kept schema source-directory paths lexical: getSchemaDir has a compatibility contract and existing symlink tests preserving lexical schema directories, so canonicalizing only schema which source paths would make the CLI inconsistent with the resolver API. Concrete template file paths remain canonical and containment-checked.

File: docs/customization.md [hunk 7]
```diff
@@ -342,7 +342,130 @@ Path: /path/to/project/openspec/schemas/my-workflow
...
+```bash
+openspec schema override spec-driven
+```
+
+This creates only:
+
+```text
+~/.local/share/openspec/schemas/spec-driven/
+└── schema.override.yaml
+```
```
[the remark this answers is not in the document]
Comment (mehdishahdoost) [reply docs/customization.md:379]: Addressed in b67442f: the directory trees now use a defined OpenSpec data directory placeholder covering XDG, the Unix and macOS default, and Windows LOCALAPPDATA.

**Comment by mehdishahdoost** [comment]
Review follow-up is pushed in b67442f and 53d329d.

Addressed:
- clarified CLI and cross-platform customization docs
- normalized overlay text boundaries and specified the exact behavior
- preserved project-schema precedence during validation
- normalized YAML-suffixed schema lookups and guarded source invariants
- preserved task-numbering checks for instruction-only overlays
- made schema which --all skip one conflicting schema while listing the usable schemas
- strengthened staging with random names, concurrent-edit detection, malformed-content rejection, and cleanup coverage
- removed duplicate validation and unreachable reporting paths
- made macOS temp-path expectations canonical and added template containment coverage

One recommendation is intentionally partial: source schema directories remain lexical to preserve the existing getSchemaDir compatibility contract. Concrete template paths are canonicalized and containment-checked.

Validation completed: lint, TypeScript build, OpenSpec strict validation, and 183 focused tests pass.

File: test/core/artifact-graph/schema-override.test.ts [hunk 8]
```diff
@@ -0,0 +1,276 @@
...
+  tasks:
+    instruction:
+      prepend: "\\n\\nPersonal preface\\n\\n"
+      append: "  \\n\\t\\n"
+`);
+
+    expect(applySchemaOverride(base, override).artifacts[2].instruction).toBe(
+      'Personal preface\n\nPackaged tasks instruction'
+    );
+  });
```
[the remark this answers is not in the document]
Comment (mehdishahdoost) [reply test/core/artifact-graph/schema-override.test.ts:113]: Addressed in 58d697e: the boundary-normalization test now includes repeated internal spaces and an internal blank line and asserts the exact composed text.

File: test/core/artifact-graph/schema-overlay.integration.test.ts [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+    const tasks = resolveSchemaTemplate('spec-driven', 'tasks.md');
+    const proposal = resolveSchemaTemplate('spec-driven', 'proposal.md');
+
+    expect(tasks.source).toBe('user');
+    expect(tasks.path).toBe(
+      fs.realpathSync.native(path.join(templatesDir, 'tasks.md'))
+    );
+    expect(fs.readFileSync(tasks.path, 'utf-8')).toContain('Personal tasks');
+    expect(proposal.source).toBe('package');
+  });
```
[the remark this answers is not in the document]
Comment (mehdishahdoost) [reply test/core/artifact-graph/schema-overlay.integration.test.ts:139]: Addressed in 58d697e: a Windows-compatible test now resolves nested/../tasks.md, verifies the user source, and compares the result with the canonical tasks.md path.

**Comment by mehdishahdoost** [comment]
Incremental CodeRabbit follow-up is pushed in 58d697e.

The initial overlay installation now commits the staged inode through a same-directory hard link. Link creation is atomic and fails with EEXIST if another process creates the destination after the initial check, so the concurrent file is preserved and the command reports the existing override instead of replacing it. Force replacement keeps its fingerprint and backup flow.

Also added the requested Windows-compatible confined alias-path test and exact internal-whitespace preservation assertion. Lint, build, strict OpenSpec validation, and all 42 affected tests pass.

**Comment by mehdishahdoost** [comment]
@coderabbitai full review

**Comment by mehdishahdoost** [comment]
@coderabbitai resume

File: docs/troubleshooting.md [hunk 10]
```diff
@@ -166,6 +166,24 @@ openspec schema init <name>         # create a custom one
...
+
+If you only need additive personal customization, use a layered override instead:
+
+```bash
+openspec schema override spec-driven
+openspec schema validate spec-driven
+openspec schema which spec-driven
+```
+
+Move your changes into `schema.override.yaml`, keep only templates you intentionally replace, and remove the complete user `schema.yaml` after preserving any custom content you still need. See [Global Overrides](customization.md#global-overrides).
```
[the remark this answers is not in the document]
Comment (mehdishahdoost) [reply docs/troubleshooting.md]: Addressed in 559fe1c: the migration now backs up and removes or renames the active complete schema first, creates the overlay second, ports supported customizations, and validates the effective result last.

File: test/cli-e2e/validate-task-numbering.test.ts [hunk 11]
```diff
@@ -198,6 +198,33 @@ describe('openspec validate checks task numbering (#1520)', () => {
...
+      { cwd: projectDir, env: { XDG_DATA_HOME: xdgDataHome } }
+    );
+
+    expect(result.exitCode).toBe(1);
+    const report = JSON.parse(result.stdout);
+    const taskIssues = report.items[0].issues.filter(
+      (issue: { path: string }) => issue.path === 'tasks.md'
+    );
+    expect(taskIssues).toHaveLength(3);
+  });
```
[the remark this answers is not in the document]
Comment (mehdishahdoost) [reply test/cli-e2e/validate-task-numbering.test.ts:244]: Addressed in 559fe1c: the numbering regression first runs schema which with the same XDG data directory and asserts the exact active user overlay path before checking numbering warnings.

File: src/commands/schema.ts [hunk 12]
```diff
@@ -633,12 +845,20 @@ export function registerSchemaCommand(program: Command): void {
...
 
         if (options?.json) {
+          const effectivePath = result.path ?? sources?.base.dir;
+          const effectiveBasePath = result.basePath ?? sources?.base.schemaPath;
           console.log(JSON.stringify({
             name,
-            path: schemaDir,
+            ...(effectivePath ? { path: effectivePath } : {}),
+            ...(effectiveBasePath ? { basePath: effectiveBasePath } : {}),
+            ...(result.overlayPath ? { overlayPath: result.overlayPath } : {}),
```
[the remark this answers is not in the document]
Comment (mehdishahdoost) [reply src/commands/schema.ts:875]: Addressed in 559fe1c: schema validate normalizes the lookup once and uses the normalized name for source resolution, effective validation, JSON, text output, and not-found messages. The extension regression now covers validation output.

File: test/core/artifact-graph/schema-override.test.ts [hunk 13]
```diff
@@ -0,0 +1,277 @@
...
+  tasks:
+    requires:
+      remove: [proposal]
+      add: []
+`);
+
+    expect(applySchemaOverride(baseSchema(), override).artifacts[2].requires).toEqual([
+      'design',
+    ]);
+  });
```
[the remark this answers is not in the document]
Comment (mehdishahdoost) [reply test/core/artifact-graph/schema-override.test.ts:194]: Addressed in 559fe1c: the additive dependency test now removes proposal, adds design as a valid non-empty operation, and asserts deterministic output.

File: src/commands/schema.ts [hunk 14]
```diff
@@ -371,6 +477,89 @@ function fingerprintDir(dir: string): string {
...
+        if ((error as NodeJS.ErrnoException).code === 'EEXIST') {
+          throw new Error(
+            `Schema override already exists at ${destinationPath}. Use --force to replace it.`
+          );
+        }
+        throw error;
+      }
+      fs.rmSync(stagingPath, { force: true });
+      return;
+    }
```
[the remark this answers is not in the document]
Comment (mehdishahdoost) [reply src/commands/schema.ts:545]: Addressed in 559fe1c: EPERM, ENOSYS, and EXDEV now fall back to exclusive write creation with the already validated staged content. EEXIST remains a friendly existing-override error, and all three fallback codes have regression coverage.

File: src/core/artifact-graph/schema.ts [hunk 15]
```diff
@@ -44,6 +64,151 @@ export function parseSchema(yamlContent: string): SchemaYaml {
...
+    ? { ...base.apply, requires: [...base.apply.requires] }
+    : undefined;
+  if (override.apply) {
+    apply ??= { requires: [] };
+    if (override.apply.requires !== undefined) {
+      apply.requires = applyCollectionOverride(
+        apply.requires,
+        override.apply.requires,
+        'apply.requires'
+      );
```
[the remark this answers is not in the document]
Comment (mehdishahdoost) [reply src/core/artifact-graph/schema.ts:182]: Addressed in 559fe1c: complete effective schema validation now checks apply.requires against artifact IDs. Add and replace overrides with an unknown prerequisite both have regression coverage.

**Comment by mehdishahdoost** [comment]
Full-review follow-up is pushed in 559fe1c.

In addition to the six inline findings, this commit handles all three review-body notes: schema file paths reuse SCHEMA_FILE_NAME, the overlay template directory is computed once, and human schema listing labels now have regression coverage.

Validation completed locally: lint, TypeScript build, strict OpenSpec validation, and 115 focused tests pass.

**Comment by mehdishahdoost** [comment]
@coderabbitai The current Merge Risk text still names two findings that are already fixed: schema which --all catches SchemaLoadError per schema and continues, and initial non-force creation uses an atomic no-replace hard link with an exclusive-write fallback. Both paths have regression coverage, all review threads are resolved, and your latest review reports no actionable comments. Please refresh the Merge Risk section to reflect the current code.
